### PR TITLE
[To be deprecated] Documentation of RDToolkit integration

### PR DESCRIPTION
**Pull request by latin-panda**
Documentation for RDToolkit integration.

Ticket: https://github.com/medic/cht-core/issues/6981
CHT Core PR: https://github.com/medic/cht-core/pull/7093
CHT Android PR: https://github.com/medic/medic-android/pull/175

### CODE REVIEW

**Comment by latin-panda** [comment]
Hi @craig-landry can you please proofread this documentation? Thanks!

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (craig-landry) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:57]: Does this image reference work? I would have expected a path of some sort to the image.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (craig-landry) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:104]: Same question about this image and if it works as expected.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (craig-landry) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:118]: This `// form` comment looks like it could use some elaboration.

**Review by craig-landry — CHANGES_REQUESTED** [review]

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:14]: Fixed

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:41]: Fixed

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:60]: Nice! Thanks!

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:118]: // The form associated with the task creation.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:57]: It turns out that the folder needs to be called the same as this file and it doesn't need path
Fixed!

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:104]: It turns out that the folder needs to be called the same as this file and it doesn't need path
Fixed!

**Comment by latin-panda** [comment]
Hi @craig-landry, great feedback! can you please take a look again? Thanks

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:18]: Out of curiosity, what happens if you have the provision and capture fields in the same form?

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:24]: It would be helpful to specify where those fields are expected to be. Are they are the top level, or adjacent to the rdtoolkit field?

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:22]: I can't think of other examples that require two appearances, so I am curious why two appearance fields are needed here? What happens if only one is set?

**Comment by craig-landry** [comment]
@latin-panda 
> can you please take a look again?

Thanks for the changes. I'm cool with those areas as they are now. I see @abbyad has some requested changes so I'll let him approve after those are resolved.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:22]: The `rdtoolkit-provision` attaches the widget and sets the scope of the widget, so we know where to display the preview and where to get the rdtoolkit exclusive fields (example: rdtoolkit_session_id field). 

The `rdtoolkit-action-btn` defines where the button to open RDToolkit app is going to display, for example we might want to display the button before a disclaimer label. If this is not provided, then the user won't be able to provide a RDTest.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:18]: Good question, if we put both widgets next to each other (to simulate pages), the pyxform return errors because the groups and fields have similar name. I guess if we merge the 2 into one group then probably we'll have 2 buttons displaying at the same time (provision and capture), where after provisioning will have to wait until the capture is ready to be able to safe the form..  but keeping both in the same form will require a bit of work 🤔

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:24]: Good catch! So, I've included the image of the xls, but definitely adding the expected level will be better.
This is fixed!

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:22]: If omitting the `rdtoolkit-action-btn` would make it not work, it would be better to make the default button included with `rdtoolkit-provision`. If now or in the future we want to separate that out, we could add a second appearance value which makes it explicit, like `rdtoolkit-action no-button`.

Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md]: Done! I've removed the class and updated documentation and code.

**Comment by latin-panda** [comment]
Hi @abbyad, thank you for the feedback! The PR has been updated, can you please review again? Thanks for your time!

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:62]: Seems like a lot of fields that need to be manually added to the XLS Form. It would have been nice for the widget to automatically add the required fields. Are they explicitly set so that the inputs can be set, and perhaps so that the output can be used in the form?

Also, if these fields persist they may be better structured as a table with columns: `Field name`, `XLSForm type`  `Description`, `Parameter type` (eg input or output), and `required`.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:101]: Same as above, would be easier to understand as a table, if these are needed to be explicitly set at all.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:68]: Same as above, would be nice for the `data` group name not to be required.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:182]: Make sure to use the ref or relref function to [avoid broken links](https://docs.communityhealthtoolkit.org/contribute/docs/style-guide/#avoid-broken-links).

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [inline content/en/apps/guides/forms/rdtoolkit-integration.md:219]: Again, worth fixing these to [avoid broken links](https://docs.communityhealthtoolkit.org/contribute/docs/style-guide/#avoid-broken-links).

**Review by abbyad — CHANGES_REQUESTED** [review]
I like the changes made as it is making it easier to use. Some comments inline, and the biggest one is about how the widget is structured. Ideally you could have multiples of any widgets on a page, which breaks if the group name is hard coded. Also, if we are hardcoding the widget name, `data` may be too generic and it appears as though it would cause problems if forms already have that field name. Some XLSForm converters actually name the top level group `data`, which make compatibility with broader XForms more difficult, but more practically, [forms in the default config use the `data` group already](https://github.com/medic/cht-core/blob/c0d452c63b9efefa5942696b184c44f3bba06e96/config/default/forms/app/pregnancy.xml#L913). 

Also, it seems as though this widget should be added to the list in to the [CHT XForms Widgets](https://docs.communityhealthtoolkit.org/apps/reference/forms/app/#cht-xform-widgets).

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:20]: Originally I didn't have required groups, but then after some conversations about making specific selectors for the xml and the limited structure for making specific selectors without recurring to ids, though that must data were expected to be in "data" group.  
I'm going to look at the options to fulfil both concerns and comeback with my findings.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:62]: Right, I'm going to switch this documentation to a table.

> It would have been nice for the widget to automatically add the required fields.

The downside I see about automatically added fields is that it isn't obvious they exits or what is group/level or how are they called when looking at the form definition (fastest way to know the actual form fields), then app devs need to take the extra step of opening the documentation or finding the form in the db, this happens "n" times. While defining the fields is a less frequent task that happens 1 time.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:182]: Nice! I'll fix it

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:20]: Thanks, it's tricky, but worth keeping it as easy as possible for app developers and also maintaining compliance and styling with XForms and XLSForms.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:62]: Ya, I see what you are saying, and that makes sense for the return values. It is in fact how ODK XLSForms have [intents that return multiple values](https://docs.getodk.org/launch-apps-from-collect/#external-apps-to-populate-multiple-fields). 

For the input parameters, you can see the way ODK handles that in their [external intent notation](https://docs.getodk.org/launch-apps-from-collect/#specifying-a-uri-as-intent-data). In the `body::intent` column they pass the URI and then the parameters with the XPath notation (eg `uuid=/myform/group/patient_uuid, ...`) or XLSForm variable notation (eg `uuid=${patient_uuid}, ...`). Using this notation makes it clean and more expected to XLSForm developers, and with the input parameters included like this we don't require unnecessary data fields either. There are some tricky aspects of how you can pass data to the XForm from XLSForm (eg like `body::intent` as used in external intents, and `instance::type`). You might find [this documentation on prefixes that add data to XForm attributes](https://xlsform.org/en/#advanced-use-and-extensibility) helpful for that. If you have a proposed format I am happy to review together and discuss before you reimplement it.

Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:62]: I'm learning this form tech on the way 😅 and not sure if I'm understanding the idea 🤔 

If using the ODK: 
As I can read the [external intent in general](https://docs.getodk.org/launch-apps-from-collect/#specifying-a-uri-as-intent-data) are available through their Android App (to be able to trigger intents) that would require users to install this extra app [ODK Collect](https://docs.getodk.org/collect-install) (+ cht-android, + rdtookit app) which will change the current flow too.

If not using that, then:
I think it's possible to create a custom column with the parameters, instead of typing these in rows then they will type them in 1 cell like this: 

[   myColumn  ]
[  rdtoolkit?filter="mal_pv"&api="https://example.com"   ]
Basically just `rdtoolkit_filter` and the `rdtoolkit_api_url` because the rest are autofilled, this solution is probably is new for us 🤔.

Because fields like `patient_uuid` still needs to be there since that's coming from the `contact` source control and managed somewhere else in cht 🤔 gotta test. Regarding the `patient_id`, `patient_name` and `rdtoolkit_session_id`, I think those should stay as fields, because they probably want to display it as read only to cross check with the dashboards or the rdtoolkit app. 

I try to imagine myself filling the form as app dev to feel the difference between solutions, however this is something I would suggest to test with users because of effort/time in development/testing/review, we'll probably get insights about other pieces of this feature too. Otherwise we can go ahead with the second solution (custom column) 🙂

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:20]: I think I've found a middle point:

We were searching by just field name prefixed by "rdtoolkit_" but there was a concern and moved it to specific selector:
Before:
 `$widget.find('input[name+="FIELD_NAME]"')`

Now it includes the form id and the group name `data`:
`$widget.find('input[name^="FORM_ID/data/FIELD_NAME]')`

[Recommended] We can relax the specific selector a bit, by looking for something that starts with the form id, omit anything in the middle (groups, or nested groups) and that ends with the field name, like a sandwich:
`$widget.find('input[name^="/FORM_ID"][name$="/FIELD_NAME]');`

Otherwise we might need to have our own group name: "rdtookit" 🙃

Note: we know what is the form id, it's variable provided by Enketo. I mean not hardcoded.

To confirm, is the sandwich solution okay? 🥪

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:62]: Sorry for the confusion! No need to use ODK Collect itself, just imitate it's technique. Most forms that work in ODK Collect would work in the CHT, but the reverse is not always true because of CHT-specific widgets (eg the contact-selector).

In terms of other input fields, they might be in the form, but not necessarily in the same group. For consistency with external intents, I'd suggest a `body::intent` column have the following:
`org.rdtoolkit.provision(filter="mal_pv", api="https://example.com", uuid=/form/inputs/patient_uuid, id=${patient_id}, name=${patient_name})`

If you want to use an `appearance` of `rdtoolkit` that would ensure that the rdtoolkit widget is used, which may make it more future proof if a generic external intent widget is added _later_ (which would be a nice feature, and help us be compatible with ODK XForms).

That makes it have the following **inputs** passed in the _parameter list_, all of which we'd generally use but in theory could be optional:
- `id` (required, only to differentiate tests)
- `name` (required, only to differentiate tests)
- `uuid` (how is this used?)
- `filter`
- `api`

And the **return** values as **fields within the same group**, all of which could be optional:
- `rdtoolkit_session_id`
- `rdtoolkit_state`
- `rdtoolkit_time_started`
- `rdtoolkit_time_resolved`

Something similar would also need to be done for the capture side.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:20]: I am not certain of the best approach for the selector... but given that we do this in other widgets there should be a way to get this to work without hardcoding the field name. Perhaps this [section of the mrdt widget](https://github.com/medic/cht-core/blob/6e82151d120e57f2332d6cc1ea78da7a5bae69f9/webapp/src/js/enketo/widgets/mrdt.js#L49-L66) can serve as an example? Given @garethbowen's [work on the mrdt widget](https://github.com/medic/cht-core/commit/50c21bf39a17a5c242492c0359db451e46002061#diff-504326969274038d01a4fffb98edf8c4905c1be1f0e6545296a9fefe77e202c9) he may be a good resource for this.

Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:20]: Looking at it, mrdt is using appearance tags as field name like: [or-appearance-mrdt-time-taken](https://github.com/medic/cht-core/blob/6e82151d120e57f2332d6cc1ea78da7a5bae69f9/webapp/src/js/enketo/widgets/mrdt.js#L61) to find and assign information into fields, but it's not really different as it has the same issue about not having specific selectors and more than one could match. 

I think the main difference with this integration is that we need to send more than 1 field and we need to receive more than 1 as well. Fields names won't be avoidable in this case as we need to make sure we're assigning the information in the right place.

I'm going to implement this solution that relaxes the selector a bit: `$widget.find('input[name^="/FORM_ID"][name$="/FIELD_NAME]');`

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:62]: Basically, support both form fields and "query string".  I'm going to do some experiments, it might take some time as I research

Comment (abbyad) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:62]: Right, and to clarify _inputs_ would only be via "query string", and _outputs_ would be form fields.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:20]: By `[name$="/FIELD_NAME]'`, are you talking about the `or-appearance-*` name, or the XLSForm field name? If it is the `or-appearance-rdtoolkit-provision` then I think that's the more standard way to do it.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:62]: From the UX/UI perspective, we still need the patient data (inputs) as fields because that's how the CHW know who they are provisioning RDTs: 
Screen Shot 2021-06-14 at 11 12 28 am 

Patient's uuid is needed for the contact source field association the patient, example: if there's a value it won't ask for the contact, if it's empty then the form ask for the contact (that's how it's working already in cht) it saves one step to the CHW:
Screen Shot 2021-06-14 at 10 02 09 pm

What I'm going to do is to leave open the possibility of receiving the contact information by query string, if this information is missing then I'll take it from the fields. I'll explain both in the docs, but the example probably more oriented to having the patient as fields since it's probably the most common scenario.  🤔

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (latin-panda) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:20]: `FIELD_NAME` is the element name in the XML tag and from the "name" column in the XLS, the standard in both.

Screen Shot 2021-06-14 at 10 10 35 pm

I think using appearance or any extra tag to identify fields is overhead, adds to complexity when building the forms.

Comment (abbyad) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:20]: Using the appearance is the convention used in XLSForm for widgets like this, so I would follow that convention for the rdtoolkit group. The inner fields which are outputs can have set field names, as done in the external intent example.

File: content/en/apps/guides/forms/rdtoolkit-integration.md
Comment (abbyad) [reply content/en/apps/guides/forms/rdtoolkit-integration.md:62]: Similar to [above](https://github.com/medic/cht-docs/pull/510#discussion_r651094219), we should be following the XLSForm convention here and having the inputs in the query string as the main way to use this widget. Generally those patient fields are already at the top level, and don't need to be repeated, which causes duplication in the resulting data as well. Being able to have the fields in the group is a nice _fallback_ if fields are missing in the query string though, but may be adding unnecessary complexity to code and documentation. Feel free to follow up offline if further clarification is needed.

**Comment by latin-panda** [comment]
Deprecated by a different approach, please refer to related issue for more details.

**Pull request closed without merging**
